torch_geometric/explainability/algo/base.py

```diff
@@ -0,0 +1,168 @@
+from abc import abstractmethod
+from typing import Callable, Optional
+
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.explainability.explanations import Explanation
+
+
+class ExplainerAlgorithm(torch.nn.Module):
+    r"""An abstract class for explainer algorithms."""
+    def __init__(self) -> None:
+        super().__init__()
+
+        self.objective: Callable[
+            [torch.Tensor, torch.Tensor, torch.Tensor, int], torch.Tensor]
+        r"""
+        This method compute the loss to be used for the explanation algorithm.
+
+        This will be set by the Explainer class. It takes the
+        python:function:`loss` and evaluate it at the against the correct
+        reference depending on the explainer mode (model vs phenomenon).
+
+        Args:
+            exp_output (torch.Tensor): the output of the explanation algorithm.
+                (e.g. the forward pass of the model with the mask applied).
+            gnn_output (torch.Tensor): the original output of the GNN.
+            target (torch.Tensor): the target of the GNN.
+            target_index (int): the index of the target to explain.
+        """
+
+    @abstractmethod
+    def loss(self, y_hat: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        r"""
+        This method compute the loss to be used for the explanation algorithm.
+        Subclasses should override this method to define their own loss.
+
+        Args:
+            y_hat (torch.Tensor): the output of the explanation algorithm.
+                (e.g. the forward pass of the model with the mask applied).
+            y (torch.Tensor): the target output.
+        """
+
+    @property
+    def accept_new_loss(self) -> bool:
+        """Whether the algorithm can accept a new loss function."""
+        return True
+
+    def set_objective(
+        self,
+        objective: Callable[[torch.Tensor, torch.Tensor, torch.Tensor, int],
+                            torch.Tensor],
+    ) -> None:
+        """Sets the loss function to be used for the explanation algorithm.
+
+        Args:
+            objective (Callable): loss function.
+        """
+        self.objective = objective
+
+    @torch.no_grad()
+    def get_initial_prediction(
+        self,
+        model: torch.nn.Module,
+        g: Data,
+        batch: Optional[torch.Tensor] = None,
+        **kwargs,
```

Comment:
I would replace those with `**kwargs` to denote the input of the model. Let's not operate expect a `Data` object here.

Reply:
Is there an advantage of using `**kwargs` to pass around the input of the model vs a `Data` object ? With `**kwargs` it seemed a bit confusing, hence the `Data` object.

Reply:
I personally think it is bad practice to use `def forward(data)` since this way it is not specified which arguments are actually used within the model. Prefer something like `def forward(x, edge_index)` (or similar) to make the inputs of the model well-defined.

Reply:
Agreed. The `Data` object also does not require those attributes to be set, so it makes it less clear to the developer what arguments are actually required.